packages/iris-grid/src/IrisGridMetricCalculator.ts

```diff
@@ -1,31 +1,101 @@
+  sorts: readonly dh.Sort[];
   reverse: boolean;
 }
 
-/**
- * Class to calculate all the metrics for a grid.
- * Call getMetrics() with the state to get metrics
- */
 export class IrisGridMetricCalculator extends GridMetricCalculator {
+  // Column widths by name to keep track of columns going in and out of viewport
+  userColumnWidthsByName: Map<ColumnName, number> = new Map();
+
+  // Cached model column names to detect when the column width map update is necessary
+  private cachedModelColumnNames: readonly ColumnName[] | undefined;
+
+  private getCachedCurrentModelColumnNames = memoizeOne(
+    (columns: readonly dh.Column[]) => columns.map(col => col.name)
+  );
+
+  private updateCalculatedColumnWidths(model: IrisGridModel): void {
+    assertNotNull(this.cachedModelColumnNames);
+    const calculatedColumnWidthsByName = new Map<ColumnName, number>();
+    this.cachedModelColumnNames.forEach((name, index) => {
+      const prevColumnWidth = this.calculatedColumnWidths.get(index);
+      if (prevColumnWidth != null) {
+        calculatedColumnWidthsByName.set(name, prevColumnWidth);
+      }
+    });
+    this.resetCalculatedColumnWidths();
+    calculatedColumnWidthsByName.forEach((width, name) => {
+      const index = model.getColumnIndexByName(name);
+      if (index != null) {
+        this.calculatedColumnWidths.set(index, width);
+      }
+    });
+    trimMap(this.calculatedColumnWidths);
+  }
+
+  /**
+   * Updates the user column widths based on the current model state
+   * @param model The current IrisGridModel
+   */
+  private updateUserColumnWidths(model: IrisGridModel): void {
+    this.userColumnWidths = new Map<ModelIndex, number>();
+    this.userColumnWidthsByName.forEach((width, name) => {
+      const modelIndex = model.getColumnIndexByName(name);
+      if (modelIndex != null) {
+        super.setColumnWidth(modelIndex, width);
+      }
+    });
+  }
+
+  /**
+   * Updates the user and calculated column widths if the model columns have changed
+   * @param model The current IrisGridModel
+   */
+  private updateColumnWidthsIfNecessary(model: IrisGridModel): void {
+    // Comparing model.columns references wouldn't work here because
+    // the reference can change in the model without the actual column definitions changing
+    if (
+      this.cachedModelColumnNames != null &&
+      !deepEqual(
+        this.getCachedCurrentModelColumnNames(model.columns),
+        this.cachedModelColumnNames
+      )
+    ) {
+      // Preserve column widths when possible to minimize visual shifts in the grid layout
+      this.updateCalculatedColumnWidths(model);
+      this.updateUserColumnWidths(model);
+    }
+    this.cachedModelColumnNames = model.columns.map(col => col.name);
+  }
+
   getGridY(state: IrisGridMetricState): number {
+    // The state here seems to be a GridMetricState with stateOverrides passed from IrisGrid in the props,
+    // not guaranteed to be IrisGridMetricState
```

Comment:
Mmmm the intent is that the `GridMetricState` combined with the `stateOverrides` _should_ produce an `IrisGridMetricState`, so it _should_ always be an `IrisGridMetricState` here... I don't know if we can do generic typing in Grid to make that enforceable. 
I guess we use the same `stateOverride` object for both `IrisGridRenderState` and `IrisGridMetricState`...
I think what I'd do is define `IrisGridStateOverride` with all the properties we set there, then `IrisGridMetricState = GridMetricState & IrisGridStateOverride`, and `IrisGridRenderState = GridRenderState & IrisGridStateOverride`. Then it should all be typed correctly?